---
title: "3_alpha_diversity"
author: "Pedro Beschoren da Costa"
date: "August 27, 2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


#General Settings

## Load libraries
[MA]: Load only packages that you'll need in this script

```{r}

#Set working directory to project directory if starting fresh
#setwd("./")
#getwd() #ok

#Handling data 
library("dplyr")
library("tibble")
library("stringr")  # to wrangle string vectors
library("tidyr")
library("ggrepel")  # to avoid legends overlapping in your plot
library("ggpubr")
library("purrr")  # has map() to select table elements

#Microbiome
library("vegan")  # for several essential statistical tests
library("forcats")
library("metagMisc")  # lets you create lists of split phyloseq objects
library("phyloseq")
library("decontam")
library("metagenomeSeq")
library("Biostrings")
library("mikropml")

#Visuals
library("ggplot2")

#stats
library("agricolae")  # includes some ANOVA post-hoc options
library("minpack.lm")  # lets you do some HSD tests, output is a nice table
library("car")  # for Levene's test

```

## Load data 
[MA] If continuing from script 2 you should have "normalization_listed" and "ps_list_CSS" in your environment
[MA] If yes, just go straight to line 68. If not, load data 
```{r}

#Load 'normalization_listed', a ps list with CSS, rarefied and non-rarefied
normalization_listed <- readRDS("./R output/01_phyloseq_objects/normalization_listed.rds")

#Get rarefied data 
#First full ps with 2 species and 2 compartments 
physeq_filtered_rarefied <- normalization_listed$physeq_filtered_rarefied

#Now load rarified data split by Species and Compartment
ps_list_rarefied <- readRDS("./R output/01_phyloseq_objects/ps_list_rarefied.rds")

```

## Settings for plots 
[MA]: Load general settings for plots (axis, treatment colors, etc)
```{r}

#Set plots theme
ggplot2::theme_set(theme_bw())

# Set default axis looks
axis_looks <- theme(axis.text.x = element_text(colour = "black", size = 9,
                                               face = "bold", angle=0, hjust=0.5))+ #text in X
  theme(axis.text.y = element_text(colour = "black", size = 11, face = "bold"))+ #text in y
  theme(axis.title=element_text(size=11, face = "bold"))+ #all texts
  theme(axis.title.y = element_text(margin = ggplot2::margin(t = 0, r = 10, b = 0, l = 0)))+
  theme(panel.grid.major = element_blank(), #no grid
        panel.grid.minor = element_blank()) #no grid

#Set colors
color_treatments <- c("#d7d7d7", "#f2c476", "#bd5a28", "#7aaf91") #"Control", "MeJA 0.1mM", "MeJA 1mM", "P. brassicae OS"
color_species <- c("#006c3a", "#309190") #at & bo, respectively
color_compartment <- c("#B28F45", "#846475") #endo & rhizo, respectively (not really used)

```



#Alpha diversity analysis
Aplha diversity will show you the quantifiable diversity within a sample - this way you compare which treatment is more or less diverse or species-rich than the other. Like in all the other steps we are evaluating here, there are countless variations and methods you could use. this is just a convenient template for you to get started.

Alpha diversity requires counts from samples with equal samples sizes. this means we cannot use the CSS transformation from the metagenomeseq package. also, heavy filtering of your data may skew some of the diversity metrics (due to the lack of rare species), so we should not over-filter the data before calculating alpha diversity.

simple Alpha diversity plots
```{r}

# as noted above, we will use rarefied data, filtered at at least 8 occurences in the dataset
# This is a very simple plot, but will serve for a visualization
plot_richness(physeq_filtered_rarefied, x = "MeJA_treatment",
              measures = c("Observed"), color = "Plant_species") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90)) +
  labs(color = "Plant species") +
  ylab("Observed number of ASVs") +
  facet_wrap(~Sample_type)

# as we could already expect, Rhizosphere is more diverse than Endosphere. For Brassica However the difference, if significant, is very small

```

##Full

### Stats
Here we check Shannon diversity indexes
```{r}
# Calculate richness for root and soil
total_diversity <- estimate_richness(physeq_filtered_rarefied) %>%
  rownames_to_column(var = "Sample")
total_diversity$Sample <- sub("X", "", total_diversity$Sample) # why that X popped up? this fix it
alpha_diversity <- column_to_rownames(total_diversity, var = "Sample") #df with only alpha metrics


#### add diversity metrics to mapping file of phyloseq objects
# we do this so we can perform anovas, acess metadat, make nicer plots, etc
merg_to_ps <- sample_data(alpha_diversity) # makes the diversity calculations  sample_data for phyloseq oject...
sample_data(physeq_filtered_rarefied) <- merge_phyloseq(physeq_filtered_rarefied, merg_to_ps) # merge the new phyloseq object with the old phyloseq object
total_diversity <- as(sample_data(physeq_filtered_rarefied), "data.frame") # forces sample data of updated phyloseq object into a dataframe


# check homogeneity of variances
leveneTest((Shannon) ~ Plant_species * MeJA_treatment * Sample_type, data = total_diversity) #[MA] not sig, ok

#Overall community [Table S4]
# two-way anova, shanon diversity index: 
tx <- with(total_diversity, interaction(MeJA_treatment, Plant_species, Sample_type)) # needed for tukey to test interaction
aovTukey <- aov(Shannon ~ tx, data = total_diversity) # needed for tukey to test interaction

#Get results from LM testing interactions in Shannon diversity 
lm <- Anova(lm((Shannon) ~ Plant_species * MeJA_treatment * Sample_type,
               data = total_diversity,
         contrasts = list(MeJA_treatment = contr.sum, Plant_species = contr.sum)), type = "2")

lm <- as.data.frame(lm) #[MA]: This is Table S4

#[MA] Get table into clipboard to paste it directly in excel
write.table(lm, "clipboard", sep = "\t", row.names = TRUE) 

#check post-hoc
HSD.test(aovTukey, "tx", group = TRUE, console = TRUE) # post-hoc

```

###Plot
```{r}

# [MA] Lets see this on a simple boxplot for Shannon diversity index
plot_alpha_Shannon <- ggplot(total_diversity,
                       aes(x = MeJA_treatment, y = Shannon, fill = MeJA_treatment)) +
  geom_boxplot() +
  theme_bw() +
  geom_jitter(data = total_diversity,
                 mapping = aes(x = MeJA_treatment, y = Shannon, fill = MeJA_treatment),
              shape = 21, size = 2, color = "black", alpha=0.8,
              position = position_jitterdodge(jitter.height = .1, jitter.width = .5))+
  labs(y = "Shannon diversity index") +
  scale_fill_manual(values = color_treatments)+
  facet_grid(Plant_species~Sample_type)+
  axis_looks+
   theme(axis.text.x = element_blank(), axis.title.x = element_blank())+
   theme(axis.ticks.x = element_blank())+
   theme(legend.position = "top")

plot_alpha_Shannon 


# [MA] save plot, [This is Figure S3]
mapply(function(x)
  ggsave(
  plot = last_plot(),
  filename =x,
  path = "./Plots/03_alpha_diversity",
  scale = 1,
  width = 20,
  height = 20,
  units = "cm",
  dpi = 300),
x=c("Shannon.svg","Shannon.png"))


```


##Species-Root Compartment

###Stats
Now that we see clear diversity effects, let's check in details to detect key pairwise differences wiht lists and loops

[MA]: For this you will use the rarefied list split by Species and Root compartment "ps_list_rarefied"

```{r}

# this will add the diversity metrics to a phyloseq object
add_diversity_to_physeq_object <- function(phyloseq_object) {

  # Calculate richness for root and soil of a sigle phyloseq object
  total_diversity <- estimate_richness(phyloseq_object) %>%
    rownames_to_column(var = "Sample")

  total_diversity$Sample <- sub("X", "", total_diversity$Sample) # why that X popped up? this fix it
  total_diversity <- column_to_rownames(total_diversity, var = "Sample")


  #### add diversity metrics to mapping file of phyloseq objects
  # we do this so we can perform anovas, acess metadat, make nicer plots, etc
  merg_to_ps <- sample_data(total_diversity) # makes the diversity calculations  sample_data for phyloseq oject...
  sample_data(phyloseq_object) <- merge_phyloseq(phyloseq_object, merg_to_ps) # merge the new phyloseq object with the old phyloseq object
  diversity <- as(sample_data(phyloseq_object), "data.frame") # forces sample data of updated phyloseq object into a dataframe

  return(diversity)
}


# runs new custom function on a list of phyloseq objects, making a list of dataframes with alpha diversity indexes
diversity_list <- lapply(ps_list_rarefied, add_diversity_to_physeq_object)


# check levenes homogeniety over the list
lapply(diversity_list, function(x) {
  leveneTest((Shannon) ~ MeJA_treatment, data = x) #ok
})

# since our initial object has been sliced into multiple parts, the anova is much simpler
# build model
aov_list <- lapply(diversity_list, function(x) {
  aov(lm((Shannon) ~ MeJA_treatment, data = x))
})

# run anova
aov_l <- lapply(aov_list, anova)
aov_l

 # check tukey
lapply(aov_list, function(x) {
  TukeyHSD(x, "MeJA_treatment")
})

#[MA]get df 

#[MA] Function to extract a cleaned ANOVA table from each element
extract_anova_table <- function(anova_result, sample_name) {
  as.data.frame(anova_result) %>%
    tibble::rownames_to_column("Source") %>%
    mutate(Sample = sample_name)}

#[MA] Apply this function to every element in the list
combined_anova <- map2_df(aov_l, names(aov_l), extract_anova_table)

#[MA] split the Sample column into Plant_species and Compartment
combined_anova <- combined_anova %>%
  separate(Sample, into = c("Plant_species", "Compartment"), sep = "\\.")

#[MA] Final combined table [This is Table S5]
print(combined_anova)

#[MA] Copy it to excel
write.table(combined_anova, "clipboard", sep = "\t", row.names = FALSE) #then copy directly in excel 


```

#Save session
```{r}

#[MA] First let's save the alpha diversity results:
alpha_diversity #full alpha diversity
diversity_list  #alpha diversity by combination of Species and Root compartment
physeq_filtered_rarefied #phyloseq with alpha diversity results in sample_Data

#[MA]Save full 
saveRDS(alpha_diversity,
        "./R output/03_alpha_diversity/alpha_diversity_full.rds") 

#[MA]Save list
save(diversity_list,
        file= "./R output/03_alpha_diversity/alpha_diversity_BySpeciesAndRootCompartment_list.RData") 

#[MA] Save ps 
saveRDS(physeq_filtered_rarefied,
        "./R output/01_phyloseq_objects/physeq_filtered_rarefied_withAlphaDiversity.rds") 

#[MA] Let's save the environment 
save.image("./R output/env.3_alpha_diversity.RData")

#[MA] Now let's clean what we don't need for next script 5-differential abundance 
rm(list = setdiff(ls(), c("normalization_listed")))


```

Alpha Diversity analysed! proceed to scrip 5_differential_abundance!


