---
title: "8_important_taxa_overrepresentation"
author: "Pedro"
date: "2/7/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

#Load libraries 

Skip if you started in the previous script 6 and just load the new packages needed for this script
```{r}
#Set working directory to project directory 
setwd("./")
getwd() #ok

#Handling data 
library("dplyr")
library("tibble")
library("stringr")  # to wrangle string vectors
library("tidyr")
library("ggrepel")  # to avoid legends overlapping in your plot
library("ggpubr")
library("purrr")  # has map() to select table elements

#Microbiome
library("vegan")  # for several essential statistical tests
library("forcats")
library("metagMisc")  # lets you create lists of split phyloseq objects
library("phyloseq")
library("decontam")
library("metagenomeSeq")
library("Biostrings")
library("mikropml")

#Visuals
library("ggplot2")

#stats
library("agricolae")  # includes some ANOVA post-hoc options
library("minpack.lm")  # lets you do some HSD tests, output is a nice table
library("car")  # for Levene's test

#New in this script
library(metacoder)
library(ggar)

```

# Overrepresentation of JA-responsive taxa
Here we take the output of script 8 that defines teh ASVs that were importnat, and then look into the taxonomies of important taxa. It tests if  these taxa are overperesented in the importnat subset, when compared to the full subset. 

This script will awnser the following question: if 6 out of 17 taxa in RF_importnat of B.Ole roots are from f__Commonadacea, and 211 out of 2453 of the taxas in B.Ole roots are from f__Commonadacea, are these proportions similar or different?

#Get data
[MA] Get a list with the names of the ASVs per subcommunity per DA, RF and NA analyses. For this you will load the results from the previous scripts 5, 6 and 7 as well as the phyloseq object that you will use for the Fisher test

```{r}

#[MA] Load ps object rarefied and split by Plant Species and Root Compartment 
ps_list_rarefied <- readRDS("./R output/01_phyloseq_objects/ps_list_rarefied.rds")

```


## DA
```{r}
#[MA] let's start with differential abundance

#[MA] load DA results
DA_results <- readRDS("./R output/05_differential_abundance/DA_ASVs.rds")

#Change the name to list 
DA_ASV_list <- DA_results

#[MA] Let's change the names 
names(DA_ASV_list) <- c("Arabidopsis_thaliana.Root", "Arabidopsis_thaliana.Soil", "Brassica_oleraceae.Root", "Brassica_oleraceae.Soil")

DA_ASV_list #ok
sapply(DA_ASV_list, length) #ok

```


```{r}


#these results are splited by stress 
lapply(DA_results$meja0.1, function(df){head(df)})
DA_results$meja1
DA_results$caterpillarOS

#let's only keep the df's
DA_results <- list(meja0.1 = DA_results$meja0.1, meja1=DA_results$meja1, caterpillarOS = DA_results$caterpillarOS)

#make a function to do the reshuffling
combine_DA_results <- function(DA_results) {
  
  # List of treatments
  treatments <- c("meja0.1", "meja1", "caterpillarOS")
  
  # Sub-community names (assumes all treatments have the same sub-community names)
  sub_communities <- names(DA_results$meja0.1)
  
  # Function to convert rownames to ASV column and add treatment name
  prepare_DA_df <- function(df, treatment) {
    df <- df %>%
      rownames_to_column(var = "ASV") %>%
      mutate(treatment = treatment)
    return(df)
  }
  
  # Initialize empty list to hold merged data per sub-community
  merged_DA_list <- setNames(vector("list", length(sub_communities)), sub_communities)
  
  # Loop through each sub-community
  for (sub_community in sub_communities) {
    
    # Combine DA results across treatments
    combined_df <- bind_rows(
      prepare_DA_df(DA_results$meja0.1[[sub_community]], "meja0.1"),
      prepare_DA_df(DA_results$meja1[[sub_community]], "meja1"),
      prepare_DA_df(DA_results$caterpillarOS[[sub_community]], "caterpillarOS")
    )
    
    # Store combined dataframe in list
    merged_DA_list[[sub_community]] <- combined_df
  }
  
  return(merged_DA_list)
}

#apply it :)
DA_list <- combine_DA_results(DA_results)
names(DA_list) <-  c("Arabidopsis_thaliana.Root", "Arabidopsis_thaliana.Soil", "Brassica_oleraceae.Root", "Brassica_oleraceae.Soil")

#get only ASV names
DA_ASV_list <- lapply(DA_list, function(df) {
  unique(df$ASV)
})

DA_ASVs

```

##NA
```{r}

#[MA] load NA results
NA_results <- readRDS("./R output/06_network_analysis/NA_ASVs.rds")

#[MA] Change names
NA_ASV_list <- NA_results
names(NA_ASV_list) <- c("Arabidopsis_thaliana.Root", "Brassica_oleraceae.Root", "Arabidopsis_thaliana.Soil",  "Brassica_oleraceae.Soil")

NA_ASV_list #ok
sapply(NA_ASV_list, length) #ok

```


```{r}



#Now 2. Network Analysis
head(NA_results)
NA_list <- split(NA_results, NA_results$Sub_community)
names(NA_list) <- c("Arabidopsis_thaliana.Root", "Arabidopsis_thaliana.Soil", "Brassica_oleraceae.Root", "Brassica_oleraceae.Soil")

#get only ASV names
NA_ASV_list <- lapply(NA_list, function(df) {
  unique(df$Row.names)  # this is where your ASVs are stored
})
names(NA_ASV_list)

```

##RF
```{r}
#[MA] Finally 3. Random Forest

#[MA] load RF results
RF_results <- readRDS("./R output/07_random_forest/RF_ASVs.rds")

#[MA] Change names
RF_ASV_list <- RF_results

names(RF_ASV_list) <- c("Arabidopsis_thaliana.Root", "Arabidopsis_thaliana.Soil", "Brassica_oleraceae.Root",  "Brassica_oleraceae.Soil")
 
RF_ASV_list #ok
sapply(RF_ASV_list, length) #ok


```

```{r}
RF_results$at_soil$combined_df

RF_list<- list(Arabidopsis_thaliana.Root = RF_results$at_root$combined_df,
                   Arabidopsis_thaliana.Soil = RF_results$at_soil$combined_df,
                   Brassica_oleraceae.Root = RF_results$bo_root$combined_df,
                   Brassica_oleraceae.Soil = RF_results$bo_soil$combined_df)

#get only ASV names
RF_ASV_list <- lapply(RF_list, function(df) {
  unique(df$ASV)  # this is where your ASVs are stored
})
names(RF_ASV_list)

#let's see what we have
DA_ASV_list
NA_ASV_list
RF_ASV_list

```

Combine it into a single list 
```{r}

#combine ASVs, you just want the unique ones to make a JA-responsive community, now it doesn't
#matter where they came from 

combined_ASV_list <- list(
  Arabidopsis_thaliana.Root = unique(c(DA_ASV_list$Arabidopsis_thaliana.Root,
                                       NA_ASV_list$Arabidopsis_thaliana.Root,
                                       RF_ASV_list$Arabidopsis_thaliana.Root)),
  
  Arabidopsis_thaliana.Soil = unique(c(DA_ASV_list$Arabidopsis_thaliana.Soil,
                                       NA_ASV_list$Arabidopsis_thaliana.Soil,
                                       RF_ASV_list$Arabidopsis_thaliana.Soil)),
  
  Brassica_oleraceae.Root = unique(c(DA_ASV_list$Brassica_oleraceae.Root,
                                     NA_ASV_list$Brassica_oleraceae.Root,
                                     RF_ASV_list$Brassica_oleraceae.Root)),
  
  Brassica_oleraceae.Soil = unique(c(DA_ASV_list$Brassica_oleraceae.Soil,
                                     NA_ASV_list$Brassica_oleraceae.Soil,
                                     RF_ASV_list$Brassica_oleraceae.Soil)))


#check how many
sapply(combined_ASV_list, length) #ok


#make one ps object with important taxa, to continue with Pedro's code 
ps_imp_taxa_l <-mapply(function (z,y)
  prune_taxa(taxa = y, x =z ),
  y = combined_ASV_list, z = ps_list_rarefied)




```


#Summary figure (Not used)

[MA] Let's explore in a pie chart how are the families of the important taxa distributed, a different type of visualization than for the heatree. For this, first we'll select those families marked as important in our study to color them, First we'll make a pie chart with those "Important ASVs" and then compare it with the Full community (Important + "unimportant" ASVs) in an example with A. thaliana Endosphere 
```{r}

#[MA]Arrange colors

#[MA]get vector for main families 
main_families <- c(
  "Burkholderiaceae", 
  "Chitinophagaceae",
  "Comamonadaceae",
  "Flavobacteriaceae",
  "Oxalobacteraceae",
  "Rhizobiaceae",
  "Sphingobacteriaceae",
  "Streptomycetaceae",
  "Xanthobacteraceae")

#[MA]Family colors matching the new logic
Family_col <- c(
    "Burkholderiaceae" = "#F7A7A4",      # Burkholderiales
    "Chitinophagaceae" = "#56B4E9",      # Chitinophagales
    "Comamonadaceae" = "#F26B6C",        # Burkholderiales - tono diferente
    "Flavobacteriaceae" = "#0072B2",     # Flavobacteriales
    "Oxalobacteraceae" = "#FEE2C7",      # Burkholderiales - otro tono diferente
    "Rhizobiaceae" = "#FED700",          # Rhizobiales - amarillo mÃ¡s suave
    "Sphingobacteriaceae" = "#009E73",   # Sphingobacteriales
    "Streptomycetaceae" = "#CC79A7",     # Streptomycetales
    "Xanthobacteraceae" = "#D9A417")      # Rhizobiales - amarillo fuerte



```

First call the main families with its colors in the important subset

```{r}

#[MA]Example for At Root (Endosphere)
combined_asvs_at_root <- unique(c(
  DA_ASV_list$Arabidopsis_thaliana.Root,
  NA_ASV_list$Arabidopsis_thaliana.Root,
  RF_ASV_list$Arabidopsis_thaliana.Root))

#[MA]subset ps
ps_at_root_important <- prune_taxa(combined_asvs_at_root, ps_list_rarefied$Arabidopsis_thaliana.Root)

#[MA]Extract taxonomy and abundance
tax_table_at_root <- as.data.frame(tax_table(ps_at_root_important)) %>%
  rownames_to_column(var = "ASV")

clean_taxonomy <- function(x) gsub("^\\w__","",x) #make function to clean names 

#[MA] apply function 
tax_table_at_root <- tax_table_at_root %>%
  mutate(Phylum = clean_taxonomy(Phylum),
         Class = clean_taxonomy(Class),
         Order = clean_taxonomy(Order),
         Family = clean_taxonomy(Family),
         Genus = clean_taxonomy(Genus))

#[MA] Summarize by Family (you could also do Phylum or Genus)
family_abundance <- as.data.frame(otu_table(ps_at_root_important)) %>%
  rownames_to_column(var = "ASV") %>%
  pivot_longer(-ASV, names_to = "Sample", values_to = "Abundance") %>%
  left_join(tax_table_at_root, by = "ASV") %>%
  group_by(Family, Order) %>%
  summarise(TotalAbundance = sum(Abundance), .groups = "drop")


#[MA] Add a new column to classify families as top families or "Other"
family_abundance.i <- family_abundance %>%
  mutate(FamilyGroup = ifelse(Family %in% main_families, Family, "Other")) %>% 
  arrange(Order, desc(TotalAbundance))

#[MA] Fix factor levels for plotting order
family_abundance.i$FamilyGroup <- factor(family_abundance.i$FamilyGroup, levels = unique(family_abundance.i$FamilyGroup))

#[MA] Add gray for others
family_colors <- c(Family_col , "Other" = "grey70")

#[MA] plot it
a.r.i <- ggplot(family_abundance.i, aes(x = "", y = TotalAbundance, fill = FamilyGroup)) +
  geom_bar(stat = "identity", width = 1, color = "white") +
  coord_polar(theta = "y") +
  theme_void() +
  scale_fill_manual(values = family_colors) +
  labs(fill = "Family", title = "Taxonomic Composition - Top Families (At Root)")


```

Now, let's do the same but with the full community for A. thaliana - Endosphere (not only the important ones)
```{r}

#[MA] Extract phyloseq object for A. thaliana root
ps_atha_root <- ps_list_rarefied$Arabidopsis_thaliana.Root

#[MA] Collapse to family level
ps_family <- tax_glom(ps_atha_root, taxrank = "Family")

#[MA] Get family-level abundances
family_abundance <- as.data.frame(otu_table(ps_family)) %>%
  rownames_to_column(var = "ASV") %>%
  left_join(as.data.frame(tax_table(ps_family)) %>%
              rownames_to_column(var = "ASV"), by = "ASV") %>%
  group_by(Family, Order) %>%
  summarise(TotalAbundance = sum(across(where(is.numeric)))) %>%
  filter(!is.na(Family), Family != "f__")

#[MA] Clean up Family names (remove f__)
clean_taxonomy #call function

#[MA] use it
family_abundance <- family_abundance %>%
  mutate(Family = clean_taxonomy(Family),
         Order = clean_taxonomy(Order))

#[MA] Assign "Other" to everything else
family_abundance <- family_abundance %>%
  mutate(FamilyGroup = ifelse(Family %in% main_families, Family, "Other"))%>% #keeping the same families as the important 
  arrange(Order, desc(TotalAbundance)) 

#[MA] Fix factor levels for plotting order
family_abundance$FamilyGroup <- factor(family_abundance$FamilyGroup,
                                       levels = unique(family_abundance$FamilyGroup))

#[MA] call colors
family_colors 

#[MA] get total ASVs
total_ASVs <- ntaxa(ps_atha_root) 

#[MA] plot it
a.r.full <- ggplot(family_abundance, aes(x = "", y = TotalAbundance, fill = FamilyGroup)) +
  geom_bar(stat = "identity", width = 1, color = "white") +
  coord_polar(theta = "y") +
  theme_void() +
  scale_fill_manual(values = family_colors ) +
  labs(fill = "Family", title = "A. thaliana Root - Full Taxonomic Composition")

ggarrange(a.r.full, a.r.i, common.legend = TRUE) #place them next to each other for comparison
#Like this it looks clear that Burkholderiaceae and Xantobacteraceae are particularly enriched in the important subset

```



# Fisher's exact test

note: There was a problem on the loop when it is called inside the function: objects "i" in the loop were literally passed as charather "i", so classes/orders could not be found. the same code works fine when hard-coded. this was a problem when using phyloseq::subset_taxa(), but not now when using phyloseq::prune_taxa() 

```{r}

#This code was written by Pedro
# source the function that calculates fisher tests for every taxonomic level. 
source("./Code/Functions/fisher_test_on_taxa_levels.R") 

# run the custom function voer 2 lists of philoseq objects, one with imporntat taxa and other with the full taxa (for every partition)
full_result_l<-mapply(function (x,y)
                   fisher_all_taxa_groups(ps_important_taxa = x, ps_all_taxa = y),
                   x = ps_imp_taxa_l,
                   y = ps_list_rarefied,
                   SIMPLIFY = FALSE)

```

[MA] Let's check all of the output
```{r}

#only use At-Endo as an example
full_result_l$Arabidopsis_thaliana.Root$fisher_result #fisher result with p values, confidence intervals and odds ratio
full_result_l$Arabidopsis_thaliana.Root$important_l #list of taxa that were important 
full_result_l$Arabidopsis_thaliana.Root$target_in_important_n #Number of times an important ASV was counted in each taxonomic group
full_result_l$Arabidopsis_thaliana.Root$target_in_all_n #Number of times an ASV was counted for each taxonomic group in the full community
full_result_l$Arabidopsis_thaliana.Root$all_taxa_in_important_n #total number of ASVs categorized as important
full_result_l$Arabidopsis_thaliana.Root$all_taxa_in_all_n #total number of ASVs in the full community

#let's make a list only with the fisher output
fisher_result_l<- list(Arabidopsis_thaliana.Root = full_result_l$Arabidopsis_thaliana.Root$fisher_result,
                       Arabidopsis_thaliana.Soil = full_result_l$Arabidopsis_thaliana.Soil$fisher_result,
                       Brassica_oleraceae.Root = full_result_l$Brassica_oleraceae.Root$fisher_result,
                       Brassica_oleraceae.Soil = full_result_l$Brassica_oleraceae.Soil$fisher_result)

```


# Extract p values and odds ratio
Now that we have large lists of fisher tests, let's fish out important metrics for visualization
```{r}

#output
fisher_result_l
                       
# non-adjusted p values for each fisher comparison
p_values <- lapply(fisher_result_l, function(x) map(x,1))

# confidence interval of odds ratio for each fisher comparison
conf_inter <- lapply(fisher_result_l, function(x) map(x,2))

# odds ratio for each fisher comparison
odds_ratio <- lapply(fisher_result_l, function(x) map(x,3))


#apply FDR correction to the p values
fdr_p_fisher_l<-lapply(fisher_result_l, function (x){
  #single df with all p values
  pvector<-do.call(rbind.data.frame, map(x,1))
  
  # adjust the p value with fdr, then save as a list
  adjusted_p<-p.adjust(p = pvector[[1]],
                          method = "fdr")%>%
                      as.list()
  # adjust names
  names(adjusted_p)<-names(x)

  return(adjusted_p)
  
})


# subset of FDR adjusted significant p values 
lapply (fdr_p_fisher_l, function (x) 
  x[x<0.05])

min3_adjustp<-str(lapply (fdr_p_fisher_l, function (x) 
  x[x<0.1]))


```

Option 2 [Ignore]
```{r}
#make fisher results into a df 

#call a function
fisher_to_df <- function(fisher_list, subcommunity_name) {
  # Convert each taxonomic level's Fisher test into a row of a dataframe
  df <- lapply(names(fisher_list), function(taxa_level) {
    fisher_test <- fisher_list[[taxa_level]]

    data.frame(
      Sub_community = subcommunity_name,
      Taxa = taxa_level,
      P_value = fisher_test$p.value,
      Odds_ratio = fisher_test$estimate,
      CI_lower = fisher_test$conf.int[1],
      CI_upper = fisher_test$conf.int[2]
    )
  }) %>%
    bind_rows()
  
  return(df)
}

# Apply to the list 
at_root_fisher_df <- fisher_to_df(fisher_result_l$Arabidopsis_thaliana.Root, "Arabidopsis_thaliana.Root")
at_soil_fisher_df <- fisher_to_df(fisher_result_l$Arabidopsis_thaliana.Soil, "Arabidopsis_thaliana.Soil")
bo_root_fisher_df <- fisher_to_df(fisher_result_l$Brassica_oleraceae.Root, "Brassica_oleraceae.Root")
bo_soil_fisher_df <- fisher_to_df(fisher_result_l$Brassica_oleraceae.Soil, "Brassica_oleraceae.Soil")

#Now get adjusted p values 
# Convert fdr_p_fisher_l into a long dataframe
fdr_p_to_df <- function(fdr_list) {
  all_rows <- lapply(names(fdr_list), function(subcom) {
    subcom_data <- fdr_list[[subcom]]
    
    # Convert each taxon into a row with its adjusted p-value
    subcom_df <- tibble(
      Sub_community = subcom,
      Taxa = names(subcom_data),
      Adjusted_p_value = unlist(subcom_data))
    
    return(subcom_df)
  })
  
  # Combine all subcommunities into a single dataframe
  #combined_df <- bind_rows(all_rows)
  return(all_rows)
}


# Apply to your list
fdr_p_df <- fdr_p_to_df(fdr_p_fisher_l)
names(fdr_p_df) <- c("Arabidopsis_thaliana.Root", "Arabidopsis_thaliana.Soil", "Brassica_oleraceae.Root", "Brassica_oleraceae.Soil")

#join dfs
# Merge adjusted p-values into the fisher results table 
at_root_fisher_with_fdr <- at_root_fisher_df %>%
  left_join(fdr_p_df$Arabidopsis_thaliana.Root, by = c("Sub_community", "Taxa"))

at_soil_fisher_with_fdr <- at_soil_fisher_df %>%
  left_join(fdr_p_df$Arabidopsis_thaliana.Soil, by = c("Sub_community", "Taxa"))

bo_soil_fisher_with_fdr <- bo_soil_fisher_df %>%
  left_join(fdr_p_df$Brassica_oleraceae.Soil, by = c("Sub_community", "Taxa"))

bo_root_fisher_with_fdr <- bo_root_fisher_df %>%
  left_join(fdr_p_df$Brassica_oleraceae.Root, by = c("Sub_community", "Taxa"))

#save as a single df for now 
fisher_with_fdr <- bind_rows(at_root_fisher_with_fdr, at_soil_fisher_with_fdr, bo_soil_fisher_with_fdr, bo_root_fisher_with_fdr)

#save this output
saveRDS(fisher_with_fdr, "./R output/08_important_taxa/full_fisher_with_fdr_df.rds")

```
[MA: For later, you could add also the number of times present in the important subset and in the full subset, with another column with the total ASVs in important and total ASVs in full]

[Continue below]


#Heat tree
metacoder heat trees are made to visualize differences in taxonomic composition, so let's incoporate p and odds ratio values into such trees. This code was written by Pedro

```{r}

fisher_to_heatTree<-function(fisher_output_l,
                             ps_important_taxa_l){
#######################################
######## make metacoder object ######## 
#######################################
imp_heat_ps<-ps_important_taxa_l

#remove unecessary taxonomic indo (dada2id, "S__" and" above_selected)
tax_table(imp_heat_ps)<-tax_table(imp_heat_ps)[,1:6]

# let's remove the "r__"ranks from the taxonomy, they can be useful but will polute our plot
tax_table(imp_heat_ps)[, colnames(tax_table(imp_heat_ps))] <- gsub(tax_table(imp_heat_ps)[, colnames(tax_table(imp_heat_ps))],     pattern = "[a-z]__", replacement = "")


# transform from phyloseq to  taxmap object
imp_heat<-parse_phyloseq(imp_heat_ps)



#######################################
######## aplly fdr to fisher ######## 
#######################################

#apply FDR correction to the p values

  #single df with all p values
  pvector<-do.call(rbind.data.frame, map(fisher_output_l,1))
  
  # adjust the p value with fdr, then save as a list
  adjusted_p<-p.adjust(p = pvector[[1]],
                          method = "fdr")%>%
                      as.list()
  # adjust names
  names(adjusted_p)<-names(fisher_output_l)

  

#this is the name that mustch match our fisher p list
taxon_id_metacoder<-lapply(imp_heat$taxa, function (x)
                                                    x$get_name())%>%
                                                    map(1)


# now turn that list into a df
taxon_id_metacoder<-do.call(rbind.data.frame, map(taxon_id_metacoder,1))

# and change column name
colnames(taxon_id_metacoder)<-"taxa_id"

# truns the p values into a dataframe
adjusted_p_input<-as.data.frame(t(as.data.frame(adjusted_p)))%>%
  rownames_to_column()

#change column names
colnames(adjusted_p_input)<-c("taxa_id", "fdr_p")


# let's remove the "r__"ranks from the taxonomy, they can be useful but will polute our plot
adjusted_p_input$taxa_id<- gsub(adjusted_p_input$taxa_id,     
                                pattern = "[a-z]__", 
                                replacement = "")

  
# now join both df
fdr_for_heatTree<-left_join(taxon_id_metacoder,
                            adjusted_p_input,
                            by = "taxa_id")



#######################################################################
################### add fold changes! ########################################
#######################################################################


# get odds ratio for each taxa
odd_list<- map(fisher_output_l,3)%>%
  map(1)

# truns the p values into a dataframe
odd_list<-as.data.frame(t(as.data.frame(odd_list)))%>%
  rownames_to_column()

#change column names
colnames(odd_list)<-c("taxa_id", "odds_ratio")

# let's remove the "r__"ranks from the taxonomy, they can be useful but will polute our plot
odd_list$taxa_id<- gsub(odd_list$taxa_id,     
                        pattern = "[a-z]__", 
                        replacement = "")


# now join both df
fisher_for_heatTree<-left_join(fdr_for_heatTree,
                            odd_list,
                            by = "taxa_id")


#NAs will crash metadore. let's change them with 0.9999 for now
fisher_for_heatTree[is.na(fisher_for_heatTree)]<-"0.99999"

#change structure to numeric
fisher_for_heatTree$fdr_p<-as.numeric(fisher_for_heatTree$fdr_p)
fisher_for_heatTree$odds_ratio<-as.numeric(fisher_for_heatTree$odds_ratio)


#now let's dim p values above 0.1, or else we can get counfounded
#fisher_for_heatTree$fdr_p[fisher_for_heatTree$fdr_p>0.1]<-1



#######################################################################
############### now plot the heat tree ############### 
#######################################################################



set.seed(1)
      output1 <- heat_tree(imp_heat,
                 node_size = fisher_for_heatTree$odds_ratio, # n_obs is a function that calculates, in this case, the number of OTUs per taxon
                 node_color = fisher_for_heatTree$fdr_p,
                 node_label = taxon_names,
                 node_size_axis_label = "Size: odds_ratio",
                 node_color_interval = c(0, 0.15),
                 edge_color_interval = c(0, 0.15),
                 node_color_range = c("darkgreen", "seagreen", "yellowgreen", "grey"),
                 edge_color_range = c("darkgreen", "seagreen", "yellowgreen", "grey"),
                 node_color_axis_label = "Color: FDR-adjusted p",
                 layout = "davidson-harel", # The primary layout algorithm
                 initial_layout = "reingold-tilford") # The layout algorithm that initializes node locations
      
      output2 <- heat_tree(imp_heat,
                 node_size = fisher_for_heatTree$odds_ratio, # 
                 node_color = fisher_for_heatTree$odds_ratio,
                 node_label = taxon_names,
                 node_size_axis_label = "Size: odds_ratio",
                 node_color_interval = c(1, 15),
                 edge_color_interval = c(1, 15),
                 node_color_range = c("grey80", "yellowgreen", "seagreen", "darkgreen"),
                 edge_color_range = c("grey80", "yellowgreen", "seagreen", "darkgreen"),
                 node_color_axis_label = "Color: Odds ratio",
                 layout = "davidson-harel", # The primary layout algorithm
                 initial_layout = "reingold-tilford") # The layout algorithm that initializes node locations

      list <- list(output=output1, df=fisher_for_heatTree)
      return(list)
}


# add the fisher values to the heat trees
output_fisher_summary_trees<-mapply(function(x,y)
    fisher_to_heatTree(x,y),
    x = fisher_result_l,
    y = ps_imp_taxa_l,
    SIMPLIFY = FALSE)


output_fisher_summary_trees$Arabidopsis_thaliana.Root$df
output_fisher_summary_trees$Arabidopsis_thaliana.Root$output

#make a single 4-pannel plot with A-D labels
output_fisher_summary_trees_plot<-ggarrange(output_fisher_summary_trees$Arabidopsis_thaliana.Root$output,
                                            output_fisher_summary_trees$Brassica_oleraceae.Root$output,
                                        output_fisher_summary_trees$Arabidopsis_thaliana.Soil$output,
                                        output_fisher_summary_trees$Brassica_oleraceae.Soil$output,
                                        common.legend = TRUE)
                                        
output_fisher_summary_trees_plot


#[MA] save plot
mapply(function(x)
  ggsave(
  plot = last_plot(),
  filename =x,
  path = "./Plots/08_important_taxa",
  scale = 1.5,
  width = 16,
  height = 16,
  units = "cm",
  dpi = 300),
x=c("heat_trees.svg","heat_trees.png"))



```

[MA] Get data into a single df just in case 
```{r}

Atha.Endo <- output_fisher_summary_trees$Arabidopsis_thaliana.Root$df
Atha.Endo <- Atha.Endo %>% mutate(Species="AT", Compartment="Endo")

Atha.Rhizo <- output_fisher_summary_trees$Arabidopsis_thaliana.Soil$df
Atha.Rhizo  <- Atha.Rhizo  %>% mutate(Species="AT", Compartment="Rhizo")

Bole.Endo <- output_fisher_summary_trees$Brassica_oleraceae.Root$df
Bole.Endo <- Bole.Endo %>% mutate(Species="BO", Compartment="Endo")

Bole.Rhizo <- output_fisher_summary_trees$Brassica_oleraceae.Soil$df
Bole.Rhizo  <- Bole.Rhizo  %>% mutate(Species="BO", Compartment="Rhizo")


all <- rbind(Atha.Endo, Atha.Rhizo, Bole.Endo, Bole.Rhizo)
all <- all %>% filter(fdr_p < 0.2)
View(all) #same info plotted in the heattree, but not hierarchical 

```
 
#Save data
```{r}

#[MA] Let's save the environment 
save.image("./R output/env.8_important_taxa_overrepresentation.RData")

#[MA] Now let's clean what we don't need for next script 7-Random Forest
rm(list = ls())

```




microbiome analysis complete! run script 9_plant_traits to evaluate plant phenotype

